components/authz-service/storage/v2/postgres/postgres.go

```diff
@@ -1267,6 +1262,105 @@ func (p *pg) ListRulesForProject(ctx context.Context, projectID string) ([]*v2.R
 	return rules, nil
 }
 
+// ApplyStagedRules begins a db transaction, locks the rule tables, moves all staged rule updates
+// and deletes into the applied rule table, and returns the database transaction. The transaction is returned
+// so that other non-database concerns can be completed before freeing the lock to avoid race conditions.
+func (p *pg) ApplyStagedRules(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := p.db.BeginTx(ctx, nil)
+	if err != nil {
+		return p.processError(err)
+	}
+
+	_, err = tx.ExecContext(ctx,
+		`LOCK TABLE iam_project_rules; LOCK TABLE iam_rule_conditions;
```

Comment:
❓Correct me if I'm wrong, but _locking the table_ means there's no way for a concurrent request to create a new rule while this is running. The alternative would be to lock existing rows in `iam_staged_project_rules`, maybe? [There's a dozen ways to do locking](https://www.postgresql.org/docs/9.4/explicit-locking.html)... but if this does the trick, I suppose it's fine.

Reply:
We don't want people to be able to update or create anything while this code is in progress.

Reply:
I'm not sure what kind of lock this executes without a mode specified. Will it allow `READ` access? We want other services (compliance) to be able to access the existing rules, no?